File: problems/daily/Java/141. Linked List Cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
       HashMap<ListNode, Boolean> map = new HashMap<>();
       while(head != null){
           if(map.containsKey(head)){
               return true;
           }
           map.put(head,true);
           head = head.next;
       }
       return false;
    }
}

//another solution
// public class Solution {
//     public boolean hasCycle(ListNode head) {
//         ListNode sh = head;
//         ListNode fh = head;
//         while (fh != null && fh.next != null) {
//             sh = sh.next;
//             fh = fh.next.next;
//             if (sh == fh) return true;
//         }
//         return false;
//     }
// }

File: problems/daily/Python3/141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        vis = set()
        cur = head

        while cur:
            if cur in vis:
                return True
            vis.add(cur)
            cur = cur.next

        return False


#another solutions


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        
        slow = fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow is fast:
                return True
        
        return False